crates/re_viewer/src/ui/blueprint_sync.rs

```diff
@@ -0,0 +1,202 @@
+use arrow2_convert::field::ArrowField;
+use re_data_store::LogDb;
+use re_log_types::{
+    Component, DataCell, DataRow, EntityPath, RowId, SerializableComponent, TimePoint,
+};
+use re_viewer_context::SpaceViewId;
+
+use crate::blueprint_components::{
+    panel::PanelState,
+    space_view::SpaceViewComponent,
+    viewport::{
+        AutoSpaceViews, SpaceViewMaximized, SpaceViewVisibility, ViewportLayout, VIEWPORT_PATH,
+    },
+};
+
+use super::{Blueprint, SpaceView, Viewport};
+
+// TODO(jleibs): Put this somewhere more global
+pub fn store_one_component<C: SerializableComponent>(
+    log_db: &mut LogDb,
+    entity_path: &EntityPath,
+    timepoint: &TimePoint,
+    component: C,
+) {
+    let mut row = DataRow::from_cells1(
+        RowId::random(),
+        entity_path.clone(),
+        timepoint.clone(),
+        1,
+        [component].as_slice(),
+    );
+    row.compute_all_size_bytes();
+
+    match log_db.entity_db.try_add_data_row(&row) {
+        Ok(()) => {}
+        Err(err) => {
+            re_log::warn_once!(
+                "Failed to store component {}.{}: {err}",
+                entity_path,
+                C::name(),
+            );
+        }
+    }
+}
+
+// Resolving and applying updates
+impl Blueprint {
+    pub fn process_updates(&self, snapshot: &Self, blueprint_db: &mut re_data_store::LogDb) {
+        // Update the panel states
+        sync_panel_expanded(
+            blueprint_db,
+            PanelState::BLUEPRINT_PANEL,
+            self.blueprint_panel_expanded,
+            snapshot.blueprint_panel_expanded,
+        );
+        sync_panel_expanded(
+            blueprint_db,
+            PanelState::SELECTION_PANEL,
+            self.selection_panel_expanded,
+            snapshot.selection_panel_expanded,
+        );
+        sync_panel_expanded(
+            blueprint_db,
+            PanelState::TIMELINE_PANEL,
+            self.time_panel_expanded,
+            snapshot.time_panel_expanded,
+        );
+
+        sync_viewport(blueprint_db, &self.viewport, &snapshot.viewport);
+
+        // Add any new or modified space views
+        for id in self.viewport.space_view_ids() {
+            if let Some(space_view) = self.viewport.space_view(id) {
+                sync_space_view(blueprint_db, space_view, snapshot.viewport.space_view(id));
+            }
+        }
+
+        // Remove any deleted space views
+        for space_view_id in snapshot.viewport.space_view_ids() {
+            if self.viewport.space_view(space_view_id).is_none() {
+                clear_space_view(blueprint_db, space_view_id);
+            }
+        }
+    }
```

Comment:
doing all this manually won't scale. I wonder if we should use something like https://crates.io/crates/diffus

Reply:
Agreed on the scaling poorly. I think there are two main approaches to consider:
 - Editing the materialized object and re-syncing, as we're doing here, in which case something like diffus seems like a good candidate.
 - Effectively making the materialized view read-only, and instead changing the places where we modify state to instead write back changes directly to the store.